galley/pkg/resource/rewrite.go

```diff
@@ -0,0 +1,44 @@
+//  Copyright 2018 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package resource
+
+import (
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"istio.io/istio/galley/pkg/common"
+)
+
+// rewrite does a generic rewrite of an unstructured custom resource.
+func rewrite(source *unstructured.Unstructured, apiVersion string) *unstructured.Unstructured {
+	modified := source.DeepCopy()
+
+	annotations := modified.GetAnnotations()
+
+	if annotations != nil {
+		delete(annotations, common.KubectlLastAppliedConfiguration)
+	} else {
+		annotations = make(map[string]string, 1)
+	}
+	annotations[common.AnnotationKeySyncedAtVersion] = source.GetResourceVersion()
+
+	modified.SetAnnotations(annotations)
+
+	modified.SetResourceVersion("")
```

Comment:
`rewrite` is also called for UPDATE. Does clearing the UID and SelfLink cause any problems with api-server?

Reply:
Yes. This is a resource coming from a different collection. The result of the rewrite is used in an different API Group/resource set. Not deleting them *will* cause issues: The target UID will definitely be different.